plugins/catalog-react/src/utils/kindFilterUtils.ts

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useApi } from '@backstage/core-plugin-api';
+import { capitalize } from '@material-ui/core';
+import useAsync from 'react-use/lib/useAsync';
+import { catalogApiRef } from '../api';
+
+/**
+ * Fetch and return all availible kinds.
+ *
+ * @public
+ */
+export function useAllKinds(): {
+  loading: boolean;
+  error?: Error;
+  allKinds: string[];
+} {
+  const catalogApi = useApi(catalogApiRef);
+
+  const {
+    error,
+    loading,
+    value: allKinds,
+  } = useAsync(async () => {
+    const items = await catalogApi
+      .getEntityFacets({ facets: ['kind'] })
+      .then(response => response.facets.kind?.map(f => f.value).sort() || []);
+    return items;
+  }, [catalogApi]);
+
+  return { loading, error, allKinds: allKinds ?? [] };
+}
+
+/**
+ * Filter and capitalize accessible kinds.
+ *
+ * @public
+ */
+export function filterAndCapitalize(
+  allKinds: string[],
+  allowedKinds?: string[],
+  forcedKinds?: string[],
+): Record<string, string> {
+  // Before allKinds is loaded, or when a kind is entered manually in the URL, selectedKind may not
+  // be present in allKinds. It should still be shown in the dropdown, but may not have the nice
+  // enforced casing from the catalog-backend. This makes a key/value record for the Select options,
+  // including selectedKind if it's unknown - but allows the selectedKind to get clobbered by the
+  // more proper catalog kind if it exists.
+  const availableKinds = allKinds
+    .concat(forcedKinds ?? [])
+    .filter(k =>
+      allowedKinds
+        ? allowedKinds.some(
+            a => a.toLocaleLowerCase('en-US') === k.toLocaleLowerCase('en-US'),
+          ) ||
+          forcedKinds?.some(
+            f => f.toLocaleLowerCase('en-US') === k.toLocaleLowerCase('en-US'),
+          )
+        : true,
+    );
+
+  const capitalizedKinds = availableKinds.sort().reduce((acc, kind) => {
+    acc[kind.toLocaleLowerCase('en-US')] = capitalize(kind);
```

Comment:
Note that the values that come out of `getEntityFacets` is already properly cased (i.e. it takes the raw values out of the entities). That ensures that things like "MyCustomKind" doesn't end up as "Mycustomkind" because the `capitalize` call can't handle it right. So the question is - I wonder if we could leverage the facets response where possible, and only capitalize things that aren't from that source.

Reply:
Yes. This is absolutely what we should do.

Reply:
As it works now, the the query parameter is honored until we actually get the correctly cased kinds. Then we try to find the query kind with the correctly cased kind and update the component. This works really great. However, useEntityListProvider.test.tsx throws some ``` Warning: An update to UserListPicker inside a test was not wrapped in act(...).``` warnings due to this extra update. Not sure how to fix it. Some hints would be appriciated.

Reply:
Yeah that's a typical thing you can see in tests when doing "reactive" things in such a way that it detects that they might not be handled properly internally.

What you do is to wrap the mutative / environment affective action inside an `act`, like for example [here](https://github.com/backstage/backstage/blob/8fe2f2f7ac30545e2d1283e0b9e48f3ec0fa8786/packages/core-components/src/components/CopyTextButton/CopyTextButton.test.tsx#L82) or [here](https://github.com/backstage/backstage/blob/8fe2f2f7ac30545e2d1283e0b9e48f3ec0fa8786/packages/core-components/src/components/FeatureDiscovery/FeatureCalloutCircular.test.tsx#L102) or [here](https://github.com/backstage/backstage/blob/8fe2f2f7ac30545e2d1283e0b9e48f3ec0fa8786/packages/core-components/src/components/TabbedLayout/RoutedTabs.test.tsx#L60).